src/serialize.rs

```diff
@@ -88,43 +88,165 @@ pub fn node_to_stream(node: &Node, f: &mut dyn io::Write) -> io::Result<()> {
+                    let start = cursor as u64;
+                    cursor += 1;
+                    let b = b[0];
+                    if b == CONS_BOX_MARKER {
+                        let index = r.len();
+                        let new_obj = ParsedTriple::Pair {
+                            start,
+                            end: 0,
+                            right_index: 0,
+                        };
+                        r.push(new_obj);
+                        op_stack.push(ParseOpRef::SaveCursor(index));
+                        op_stack.push(ParseOpRef::ParseObj);
+                        op_stack.push(ParseOpRef::SaveIndex(index));
+                        op_stack.push(ParseOpRef::ParseObj);
+                    } else {
+                        let (start, end, atom_offset) = {
+                            if b <= MAX_SINGLE_BYTE {
+                                (start, start + 1, 0)
+                            } else {
+                                let (atom_offset, atom_size) = decode_size(f, b)?;
+                                skip_bytes(f, atom_size)?;
+                                let end = start + (atom_offset as u64) + (atom_size as u64);
+                                (start, end, atom_offset as u32)
+                            }
+                        };
+                        let new_obj = ParsedTriple::Atom {
+                            start,
+                            end,
+                            atom_offset,
+                        };
+                        cursor = end;
+                        r.push(new_obj);
+                    }
+                }
+                ParseOpRef::SaveCursor(index) => {
+                    if let ParsedTriple::Pair {
+                        start,
+                        end: _,
+                        right_index,
+                    } = r[index]
+                    {
+                        r[index] = ParsedTriple::Pair {
+                            start,
+                            end: cursor,
+                            right_index,
+                        };
+                    }
+                }
+                ParseOpRef::SaveIndex(index) => {
+                    if let ParsedTriple::Pair {
+                        start,
+                        end,
+                        right_index: _,
+                    } = r[index]
+                    {
+                        r[index] = ParsedTriple::Pair {
+                            start,
+                            end,
+                            right_index: r.len() as u32,
+                        };
+                    }
+                }
+            },
+        }
+    }
+    Ok(r)
+}
+
 /// decode the length prefix for an atom. Atoms whose value fit in 7 bits
 /// don't have a length prefix, so those should be handled specially and
 /// never passed to this function.
-fn decode_size(f: &mut dyn io::Read, initial_b: u8) -> io::Result<u64> {
```

Comment:
I missed that this turned into `dyn` at some point. changing it back seems like a good idea in itself